### PR TITLE
Handle careless use in bracket

### LINKED ISSUE #4960: Release in bracket does not get eval'd with careless use action
I was preparing a ZIO workshop at Pirum (YAY 🎉) when this broke my expectations:
```scala
ZIO.bracket(putStrLn("Start"))(_ => putStrLn("End"))(_ => throw new RuntimeException("oh noes")) 
```
... did not print `End`. I think it should, the runtime loop has logic to handle that case.

### PR DESCRIPTION
**Pull request by regiskuckaertz**
Closes #4960

### CODE REVIEW

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 1]
```diff
@@ -2331,7 +2331,7 @@ object ZIO extends ZIOCompanionPlatformSpecific {
   ): ZIO[R, E, B] =
     ZIO.uninterruptibleMask[R, E, B](restore =>
       acquire.flatMap(ZIOFn(traceAs = use) { a =>
-        restore(use(a)).run.flatMap(ZIOFn(traceAs = release) { e =>
+        restore(ZIO.effectSuspendTotal(use(a))).run.flatMap(ZIOFn(traceAs = release) { e =>
           release(a, e).foldCauseM(
```
Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/ZIO.scala:2335]: Should this also be wrapped?

**Review by regiskuckaertz — COMMENTED** [review]
Just left one question. Cannot hurt right?

**Comment by regiskuckaertz** [comment]
Looks like spurious failures

**Comment by regiskuckaertz** [comment]
Not spurious after all. @adamgfraser  looks like I plugged a leak in the stack tracking stuff. Ready for review sir.

File: core/shared/src/main/scala/zio/ZIO.scala [hunk 2]
```diff
@@ -2331,11 +2331,13 @@ object ZIO extends ZIOCompanionPlatformSpecific {
   ): ZIO[R, E, B] =
     ZIO.uninterruptibleMask[R, E, B](restore =>
       acquire.flatMap(ZIOFn(traceAs = use) { a =>
-        restore(use(a)).run.flatMap(ZIOFn(traceAs = release) { e =>
-          release(a, e).foldCauseM(
-            cause2 => ZIO.halt(e.fold(_ ++ cause2, _ => cause2)),
-            _ => ZIO.done(e)
-          )
+        restore(ZIO.effectSuspendTotal(use(a))).run.flatMap(ZIOFn(traceAs = release) { e =>
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/ZIO.scala:2334]: I think we want `restore` around `use` instead of `effectSuspendTotal`.

Comment (regiskuckaertz) [inline core/shared/src/main/scala/zio/ZIO.scala:2334]:
```suggestion
        ZIO.effectSuspendTotal(restore(use(a))).run.flatMap(ZIOFn(traceAs = release) { e =>
```
Yes, thank you.

**Review by adamgfraser — APPROVED** [review]

**Pull request merged**